### PR TITLE
Add support for TSDB selector in querier

### PR DESCRIPTION
**Pull request by fpetkovski**
This PR allows using the query distributed mode against a set of multi-tenant receivers as described in the [original proposal](https://github.com/thanos-io/thanos/blob/main/docs/proposals-done/202301-distributed-query-execution.md#distributed-execution-against-receive-components).

The feature is enabled by a selector.relabel-config flag in the Query component which allows it to select a subset of TSDBs to query based on their external labels.

The credit should go to @sharadgaur since he originally did the implementation in https://github.com/thanos-io/thanos/pull/6122. The PR went stale and this is just tying up that loose end.

* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

### CODE REVIEW

**Review by MichaHoffmann — DISMISSED** [review]
lgtm!

**Comment by MichaHoffmann** [comment]
I think LabelNames and LabelValues need similar treatment like Series

**Comment by fpetkovski** [comment]
Sounds good. I've also added some more tests for generating matchers for label sets.

**Comment by fpetkovski** [comment]
Any thoughts @saswatamcode @GiedriusS ? There seems to be some overlap with tenancy work, but I think this is more generic and useful for users that do not need such rigid enforcement.

**Review by saswatamcode — DISMISSED** [review]
LGTM! I think this is super flexible. Would be great to follow up with labelnames/values in the future!

File: docs/components/query.md [hunk 1]
```diff
@@ -460,6 +460,20 @@ Flags:
...
+                                 flag (mutually exclusive). Content of
+                                 YAML file that contains relabeling
+                                 configuration that allows selecting
+                                 blocks. It follows native Prometheus
+                                 relabel-config syntax. See format details:
+                                 https://prometheus.io/docs/prometheus/latest/configuration/configuration/#relabel_config
+      --selector.relabel-config-file=<file-path>
+                                 Path to YAML file that contains relabeling
+                                 configuration that allows selecting
+                                 blocks. It follows native Prometheus
```
Comment (GiedriusS) [inline docs/components/query.md:474]: Selecting blocks? :smile:

**Review by MichaHoffmann — DISMISSED** [review]

File: docs/components/query.md (hunk 1 above)
Comment (fpetkovski) [reply docs/components/query.md]: I updated the description to make it more clear.

**Review by MichaHoffmann — APPROVED** [review]

**Comment by yeya24** [comment]
I am trying to understand how this feature is used. You need to create different Querier deployments with different relabel configs if you want to be able to query those tenants differently?

**Comment by fpetkovski** [comment]
@yeya24 yes that is exactly correct.

**Pull request merged**